ctags/main/routines.c

```diff
@@ -0,0 +1,788 @@
+#endif
+#ifdef NEED_PROTO_LSTAT
+extern int lstat (const char *, struct stat *);
+#endif
+
+#if 0
+static boolean createTagsForEntry (const char *const entryName);
+#endif
+
+
+/*
+*   FUNCTION DEFINITIONS
+*/
+
+
+extern void setExecutableName (const char *const path)
+{
+	ExecutableProgram = path;
+	ExecutableName = baseFilename (path);
+}
+
+extern const char *getExecutableName (void)
+{
+	return ExecutableName;
+}
+
+
+extern void *eMalloc (const size_t size)
+{
+	void *buffer = g_malloc (size);
+
+	if (buffer == NULL)
+		error (FATAL, "out of memory");
+
+	return buffer;
+}
+
+extern void *eCalloc (const size_t count, const size_t size)
+{
+	void *buffer = calloc (count, size);
+
+	if (buffer == NULL)
+		error (FATAL, "out of memory");
+
+	return buffer;
+}
+
+extern void *eRealloc (void *const ptr, const size_t size)
+{
+	void *buffer;
+	if (ptr == NULL)
+		buffer = eMalloc (size);
+	else
+	{
+		buffer = g_realloc (ptr, size);
+		if (buffer == NULL)
+			error (FATAL, "out of memory");
+	}
+	return buffer;
+}
+
+extern void eFree (void *const ptr)
+{
+	if (ptr != NULL)
+		free (ptr);
+}
+
+#ifndef HAVE_STRSTR
+extern char* strstr (const char *str, const char *substr)
+{
+	const size_t length = strlen (substr);
+	const char *match = NULL;
+	const char *p;
+
+	for (p = str  ;  *p != '\0'  &&  match == NULL  ;  ++p)
+		if (strncmp (p, substr, length) == 0)
+			match = p;
+	return (char*) match;
+}
+#endif
+
+extern char* eStrdup (const char* str)
+{
+	char* result = xMalloc (strlen (str) + 1, char);
+	strcpy (result, str);
+	return result;
+}
+
+extern void toLowerString (char* str)
+{
+	while (*str != '\0')
+	{
+		*str = tolower ((int) *str);
+		++str;
+	}
+}
+
+extern void toUpperString (char* str)
+{
+	while (*str != '\0')
+	{
+		*str = toupper ((int) *str);
+		++str;
+	}
+}
+
+/*  Newly allocated string containing lower case conversion of a string.
+ */
+extern char* newLowerString (const char* str)
+{
+	char* const result = xMalloc (strlen (str) + 1, char);
+	int i = 0;
+	do
+		result [i] = tolower ((int) str [i]);
+	while (str [i++] != '\0');
+	return result;
+}
+
+/*  Newly allocated string containing upper case conversion of a string.
+ */
+extern char* newUpperString (const char* str)
+{
+	char* const result = xMalloc (strlen (str) + 1, char);
+	int i = 0;
+	do
+		result [i] = toupper ((int) str [i]);
+	while (str [i++] != '\0');
+	return result;
+}
+
+
+#if 0
+static void setCurrentDirectory (void)
+{
+	char* const cwd = getcwd (NULL, PATH_MAX);
+	CurrentDirectory = xMalloc (strlen (cwd) + 2, char);
+	if (cwd [strlen (cwd) - (size_t) 1] == PATH_SEPARATOR)
+		strcpy (CurrentDirectory, cwd);
+	else
+		sprintf (CurrentDirectory, "%s%c", cwd, OUTPUT_PATH_SEPARATOR);
+	free (cwd);
+}
+#endif
+
+
+extern boolean doesFileExist (const char *const fileName)
+{
+	GStatBuf fileStatus;
+
+	return (boolean) (g_stat (fileName, &fileStatus) == 0);
+}
+
+extern boolean isRecursiveLink (const char* const dirName)
```

Comment:
this was #if0'd, and the impl in UCTags is different (but this one matches the old one we had, so I guess it's expected)